### PR TITLE
Fix a memory misalignment in topk operator

### PR DESCRIPTION
**Pull request by apeforest**
## Description ##
Current memory alignment in topk operator is incorrect if index_t is using int64_t. This PR fixes the potential issue. It partially fixes https://github.com/apache/incubator-mxnet/issues/15703

The PR also fixes an incorrect data type usage in mshadow/tensor.h
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
Bug fix

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by apeforest** [comment]
@access2rohit @ChaiBapchya Please also help review. Thanks

File: 3rdparty/mshadow/mshadow/tensor.h [hunk 1]
```diff
@@ -69,15 +69,15 @@ struct Shape {
    * \param idx dimension index
    * \return the corresponding dimension size
    */
-  MSHADOW_XINLINE index_t &operator[](index_t idx) {
+  MSHADOW_XINLINE index_t &operator[](int idx) {
```
Comment (access2rohit) [inline 3rdparty/mshadow/mshadow/tensor.h]: Can you describe this change a bit more in detail ? Why is this required ?

File: src/operator/tensor/ordering_op-inl.h [hunk 2]
```diff
@@ -414,30 +414,23 @@ void TopKImpl(const RunContext &ctx,
...
                                     + PadBytes(sizeof(index_t) * src.Size(), alignment);
   if (param.ret_typ == topk_enum::kReturnMask) {
-    workspace_size += PadBytes(sizeof(int) * batch_size * k, alignment);
+    workspace_size += PadBytes(sizeof(index_t) * batch_size * k, alignment);
   }
   workspace = resource.get_space_typed<xpu, 1, char>(Shape1(workspace_size), s);
   char* workspace_curr_ptr = workspace.dptr_;
   sorted_dat = Tensor<xpu, 1, DType>(reinterpret_cast<DType*>(workspace_curr_ptr),
-                                      Shape1(src.Size()), s);  // contain sorted dat
+      Shape1(src.Size()), s);  // contain sorted dat
```
Comment (access2rohit) [inline src/operator/tensor/ordering_op-inl.h]: nit: indentation

File: src/operator/tensor/ordering_op-inl.h [hunk 3]
```diff
@@ -414,30 +414,23 @@ void TopKImpl(const RunContext &ctx,
...
   workspace_curr_ptr += PadBytes(sizeof(DType) * src.Size(), alignment);
   indices = Tensor<xpu, 1, index_t>(reinterpret_cast<index_t*>(workspace_curr_ptr),
-                                Shape1(src.Size()), s);  // indices in the original matrix
+      Shape1(src.Size()), s);  // indices in the original matrix
```
Comment (access2rohit) [inline src/operator/tensor/ordering_op-inl.h]: nit: indentation ?

File: src/operator/tensor/ordering_op-inl.h (hunk 2 above)
Comment (apeforest) [reply src/operator/tensor/ordering_op-inl.h]: it's 4 space continuation indent set by Google Style Guide

File: 3rdparty/mshadow/mshadow/tensor.h (hunk 1 above)
Comment (apeforest) [reply 3rdparty/mshadow/mshadow/tensor.h]: This is not directly fixing this bug. However, while checking the tensor struct, I noticed this data type should be int because it is indexing the dimension (which is set to int)

File: src/operator/tensor/ordering_op-inl.h [hunk 4]
```diff
@@ -414,30 +414,28 @@ void TopKImpl(const RunContext &ctx,
...
-  temp_size = std::max(temp_size,
-    mxnet::op::SortByKeyWorkspaceSize<int, int, xpu>(src.Size()));
-  temp_size = std::max(temp_size,
-    mxnet::op::SortByKeyWorkspaceSize<int, DType, xpu>(src.Size()));
-  temp_size = std::max(temp_size,
-    mxnet::op::SortByKeyWorkspaceSize<DType, int, xpu>(src.Size()));
+  // Temp space needed by the full sorts.
+  size_t temp_size = std::max(
+      mxnet::op::SortByKeyWorkspaceSize<index_t, DType, xpu>(src.Size()),
+      mxnet::op::SortByKeyWorkspaceSize<DType, index_t, xpu>(src.Size()));
```
Comment (sxjscience) [inline src/operator/tensor/ordering_op-inl.h]: I think we should also include `mxnet::op::SortByKeyWorkspaceSize<index_t, index_t, xpu>(src.Size())`.

File: 3rdparty/mshadow/mshadow/tensor.h [hunk 5]
```diff
@@ -69,15 +69,15 @@ struct Shape {
...
     return shape_[idx];
   }
   /*!
    * \brief get corresponding index
    * \param idx dimension index
    * \return the corresponding dimension size
    */
-  MSHADOW_XINLINE const index_t &operator[](index_t idx) const {
+  MSHADOW_XINLINE const index_t &operator[](int idx) const {
```
Comment (sxjscience) [inline 3rdparty/mshadow/mshadow/tensor.h]: There are other places in the same file that assumes idx have index_t type. we need to fix all of them.

Comment (apeforest) [reply 3rdparty/mshadow/mshadow/tensor.h]: Done

File: src/operator/tensor/ordering_op-inl.h (hunk 4 above)
Comment (apeforest) [reply src/operator/tensor/ordering_op-inl.h]: Done

**Comment by TaoLv** [comment]
Thank you for the fix, @apeforest . Will this be picked to the v1.5.x branch?

**Comment by apeforest** [comment]
@TaoLv I have cherry-picked it to v1.5.x (commit hash: 42746bc73e8bcb75bfcadd1398e6f71bc170fa10)

Somehow the v1.5.x branch was not PR protected so when I push it to origin it got merged automatically. Please let me know if this is okay. Thanks.

**Comment by TaoLv** [comment]
@apeforest  Could you please open a dummy PR against the v1.5.x branch to make sure the branch can properly pass the CI?

**Comment by apeforest** [comment]
See https://github.com/apache/incubator-mxnet/pull/15999
